### PR TITLE
Remove use of lists in the syntactic classifier

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/52662.  Will be made non-draft once that goes in.

This will be easier to review with whitespace changes off.

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs [hunk 1]
```diff
@@ -17,46 +14,19 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Classification
...
                 new ClassificationTag(typeMap.GetClassificationType(classifiedSpan.ClassificationType)));
         }
 
-        public static List<ITagSpan<IClassificationTag>> ConvertAndReturnList(ClassificationTypeMap typeMap, ITextSnapshot snapshot, List<ClassifiedSpan> classifiedSpans)
+        public static List<ITagSpan<IClassificationTag>> Convert(ClassificationTypeMap typeMap, ITextSnapshot snapshot, ArrayBuilder<ClassifiedSpan> classifiedSpans)
         {
-            var result = new List<ITagSpan<IClassificationTag>>();
-            Convert(typeMap, snapshot, classifiedSpans, result.Add);
-            ReturnClassifiedSpanList(classifiedSpans);
+            var result = new List<ITagSpan<IClassificationTag>>(capacity: classifiedSpans.Count);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs:26]: this value is passed back out to callers of ITagger.GetTags (which returns an enumerable.  In this case, it's fine to realize that as a List.

File: src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs [hunk 2]
```diff
@@ -220,7 +221,15 @@ private void OnSubjectBufferChanged(object? sender, TextContentChangedEventArgs
...
-                EnqueueWorkIfThisDocument(args.NewActiveContextDocumentId);
+                if (_workspace == null)
+                    return;
+
+                var documentId = args.NewActiveContextDocumentId;
+                var bufferDocumentId = _workspace.GetDocumentIdInCurrentContext(_subjectBuffer.AsTextContainer());
+                if (bufferDocumentId != documentId)
+                    return;
+
+                _workQueue.AddWork(_subjectBuffer.CurrentSnapshot);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs:232]: request from previous PR.

**Comment by CyrusNajmabadi** [comment]
This is ready for review.

File: src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs (hunk 1 above)
Comment (dibarbet) [reply src/EditorFeatures/Core/Implementation/Classification/ClassificationUtilities.cs:26]: I presume we're avoiding linq here for perf?

File: src/Tools/ExternalAccess/FSharp/Internal/Classification/FSharpClassificationService.cs [hunk 3]
```diff
@@ -27,27 +29,33 @@ public FSharpClassificationService(IFSharpClassificationService service)
             _service = service;
         }
 
-        public void AddLexicalClassifications(SourceText text, TextSpan textSpan, List<ClassifiedSpan> result, CancellationToken cancellationToken)
+        public void AddLexicalClassifications(SourceText text, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
         {
-            _service.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+            using var _ = s_listPool.GetPooledObject(out var list);
+            _service.AddLexicalClassifications(text, textSpan, list, cancellationToken);
```
Comment (dibarbet) [inline src/Tools/ExternalAccess/FSharp/Internal/Classification/FSharpClassificationService.cs:35]: Ideally F# should update their interface to take in an arraybuilder right?   Just want to check that they know

**Review by dibarbet — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
> I presume we're avoiding linq here for perf?

in general, yes.

File: src/Tools/ExternalAccess/FSharp/Internal/Classification/FSharpClassificationService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Tools/ExternalAccess/FSharp/Internal/Classification/FSharpClassificationService.cs:35]: Yup.  But we leave that up to them (though it's totally fine for us to notify them).  Effectively, they can choose to augment their api and do whatever they need, without affecting us.  it's the benefit of the EA pattern here.

**Pull request merged**